core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java

```diff
@@ -0,0 +1,315 @@
+    }
+    if (readAborted) {
+      throw new IOException(readException);
+    }
+    if (isEndOfStream()) {
+      return -1;
+    }
+    len = Math.min(len, activeBuffer.remaining());
+    activeBuffer.get(b, offset, len);
+
+    if (activeBuffer.remaining() <= readAheadThresholdInBytes && !readAheadBuffer.hasRemaining()) {
+      readAsync(readAheadBuffer);
+    }
+    if (!activeBuffer.hasRemaining()) {
+      ByteBuffer temp = activeBuffer;
+      activeBuffer = readAheadBuffer;
+      readAheadBuffer = temp;
+    }
+    return len;
+  }
+
+  @Override
+  public int available() throws IOException {
+    stateChangeLock.lock();
+    // Make sure we have no integer overflow.
+    int val = (int) Math.min((long) Integer.MAX_VALUE,
+            (long) activeBuffer.remaining() + readAheadBuffer.remaining());
+    stateChangeLock.unlock();
+    return val;
+  }
+
+  @Override
+  public long skip(long n) throws IOException {
+    if (n <= 0L) {
+      return 0L;
+    }
+    stateChangeLock.lock();
+    long skipped;
+    try {
+      skipped = skipInternal(n);
+    } finally {
+      stateChangeLock.unlock();
+    }
+    return skipped;
+  }
+
+  /**
+   * Internal skip function which should be called only from skip() api. The assumption is that
+   * the stateChangeLock is already acquired in the caller before calling this function.
+   */
+  private long skipInternal(long n) throws IOException {
+    assert (stateChangeLock.isLocked());
+    if (readInProgress) {
+      waitForAsyncReadComplete();
+    }
+    if (available() >= n) {
+      // we can skip from the internal buffers
+      int toSkip = (int)n;
+      if (toSkip <= activeBuffer.remaining()) {
+        // Only skipping from active buffer is sufficient
+        activeBuffer.position(toSkip + activeBuffer.position());
+        return n;
+      }
+      // We need to skip from both active buffer and read ahead buffer
+      toSkip -= activeBuffer.remaining();
+      activeBuffer.position(0);
+      activeBuffer.flip();
+      readAheadBuffer.position(toSkip + readAheadBuffer.position());
+      // flip the active and read ahead buffer
+      ByteBuffer temp = activeBuffer;
+      activeBuffer = readAheadBuffer;
+      readAheadBuffer = temp;
+      readAsync(readAheadBuffer);
+      return n;
+    }
+    int skippedBytes = available();
+    long toSkip = n - skippedBytes;
+    activeBuffer.position(0);
+    activeBuffer.flip();
+    readAheadBuffer.position(0);
+    readAheadBuffer.flip();
+    long skippedFromInputStream = underlyingInputStream.skip(toSkip);
+    readAsync(activeBuffer);
+    return skippedBytes + skippedFromInputStream;
+  }
+
+  @Override
+  public void close() throws IOException {
+    executorService.shutdown();
+    try {
+      executorService.awaitTermination(10, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+    }
+    underlyingInputStream.close();
+    stateChangeLock.lock();
+    try {
+      StorageUtils.dispose(activeBuffer);
```

Comment:
You are right, but this is being done so as to release the memory as soon as we close the inputstream otherwise, it will be freed only when GC is kicked in and in case we create too many input stream and GC does not kick in, we will have some memory leak.

Reply:
`StorageUtils.dispose` is no-op for HeapByteBuffer

Reply:
Good point, my bad. I was under impression that I am allocating a direct byte buffer instead. Fixed.